src/components/application_manager/src/app_service_manager.cc

```diff
@@ -528,4 +532,256 @@ void AppServiceManager::AppServiceUpdated(
+              current_request_[correlation_id].origin_connection_key;
+          smart_objects::SmartObjectSPtr result =
+              std::make_shared<smart_objects::SmartObject>(
+                  current_request_[correlation_id].message);
+          request_queue_.erase(correlation_id);
+          current_request_.erase(correlation_id);
+          messages_using_core_.insert(correlation_id);
+          RemoveTimeoutTimer(correlation_id);
+          app_manager_.GetRPCService().ManageMobileCommand(
+              result, commands::Command::SOURCE_MOBILE);
+        } else {
+          LOG4CXX_DEBUG(logger_,
+                        "Forwarding request to service with app id "
+                            << current_request_[correlation_id]
+                                   .destination_connection_key);
+          smart_objects::SmartObjectSPtr result =
+              std::make_shared<smart_objects::SmartObject>(
+                  current_request_[correlation_id].message);
+          AddTimeoutTimer(correlation_id);
+          app_manager_.GetRPCService().SendMessageToMobile(result);
+        }
+        return true;
+      }
+
+      LOG4CXX_DEBUG(logger_, "Valid RPC passing response");
+
+      if (current_request_.find(correlation_id) != current_request_.end()) {
+        LOG4CXX_DEBUG(
+            logger_,
+            "Forwarding response to mobile app "
+                << current_request_[correlation_id].origin_connection_key);
+        rpc_message[strings::params][strings::connection_key] =
+            current_request_[correlation_id].origin_connection_key;
+        smart_objects::SmartObjectSPtr result =
+            std::make_shared<smart_objects::SmartObject>(rpc_message);
+
+        request_queue_.erase(correlation_id);
+        current_request_.erase(correlation_id);
+        RemoveTimeoutTimer(correlation_id);
+        app_manager_.GetRPCService().SendMessageToMobile(result);
+        return true;
+      }
+
+      return false;
+    } break;
+  }
+
+  return false;
+}
+
+bool RPCPassingHandler::CycleThroughRPCPassingRequests(
+    uint32_t correlation_id) {
+  LOG4CXX_DEBUG(logger_, "Cycling through pass through requests");
+  if (request_queue_.find(correlation_id) == request_queue_.end()) {
+    LOG4CXX_ERROR(logger_, "Correlation id does NOT exist in map");
+    return false;
+  }
+
+  current_request_[correlation_id] = request_queue_[correlation_id].front();
+  request_queue_[correlation_id].pop();
+
+  MessageHelper::PrintSmartObject(current_request_[correlation_id].message);
+  return true;
+}
+
+void RPCPassingHandler::OnPassThroughRequestTimeout() {
+  LOG4CXX_DEBUG(logger_, "RPC Passing request timeout");
+  auto timeout_entry = timeout_queue_.front();
+  uint32_t correlation_id = std::get<1>(timeout_entry);
+  timeout_queue_.pop();
+  LOG4CXX_DEBUG(logger_,
+                "Cycling through requests for correlation_id "
+                    << correlation_id);
+  CycleThroughRPCPassingRequests(correlation_id);
```

Comment:
This function does not send the next message, what triggers the next attempt?

Reply:
`CycleThroughRPCPassingRequests` should probably just contain the logic for sending the next request in the list

Reply:
First comment was fixed 👍